src/core/api/message_api.cpp

```diff
@@ -176,4 +178,15 @@ void otMessageGetBufferInfo(otInstance *aInstance, otBufferInfo *aBufferInfo)
 
     aInstance->mThreadNetif.GetCoap().GetRequestMessages().GetInfo(aBufferInfo->mCoapMessages,
                                                                    aBufferInfo->mCoapBuffers);
+    aInstance->mThreadNetif.GetCoap().GetCachedResponses().GetInfo(messages, buffers);
+    aBufferInfo->mCoapMessages += messages;
+    aBufferInfo->mCoapBuffers += buffers;
+
+#if OPENTHREAD_ENABLE_APPLICATION_COAP
+    aInstance->mApplicationCoap.GetRequestMessages().GetInfo(aBufferInfo->mApplicationCoapMessages,
+                                                             aBufferInfo->mApplicationCoapBuffers);
+    aInstance->mApplicationCoap.GetCachedResponses().GetInfo(messages, buffers);
+    aBufferInfo->mApplicationCoapMessages += messages;
+    aBufferInfo->mApplicationCoapBuffers += buffers;
```

Comment:
Set these fields to zero if `OPENTHREAD_ENABLE_APPLICATION_COAP` is 0?

Reply:
Done. Also I've noticed there's a similar PR for secure coap (#2227), which should probably do the same for consistency.